vesc_hw_interface/src/vesc_step_difference.cpp

```diff
@@ -0,0 +1,123 @@
+
+#include "vesc_hw_interface/vesc_step_difference.h"
+namespace vesc_step_difference
+{
+VescStepDifference::VescStepDifference()
+{
+}
+
+VescStepDifference::~VescStepDifference()
+{
+}
+
+void VescStepDifference::enableSmooth(double control_rate, double max_sampling_time, int max_step_diff)
+{
+  enable_smooth_ = true;
+  step_diff_vw_max_step_ = max_step_diff;
+  step_diff_vw_max_window_size_ = std::max(1, static_cast<int>(control_rate * max_sampling_time));
+  step_diff_log_.resize(step_diff_vw_max_window_size_ + 1);
+  return;
+}
+
+double VescStepDifference::getStepDifference(const double step_in, bool reset)
+{
+  double output;
+  if (enable_smooth_)
+  {
+    output = this->stepDifferenceVariableWindow(step_in, reset);
+  }
+  else
+  {
+    output = this->stepDifferenceRaw(step_in, reset);
+  }
+  return output;
+}
+
+/**
+ * Function called when the filter is disabled
+ *
+ * Outputs the difference of the counter value from the previous value.
+ * This function is used when no filter is used.
+ *
+ * @step_in Enter the current counter value
+ * @reset Used for initialization.
+ * @return Returns the difference from the previous value.
+ */
+double VescStepDifference::stepDifferenceRaw(const double step_in, bool reset)
+{
+  if (reset)
+  {
+    step_diff_raw_prev_ = static_cast<int16_t>(step_in);
+    return 0.0;
+  }
+  int16_t step_diff = static_cast<int16_t>(step_in) - step_diff_raw_prev_;
+  double output = static_cast<double>(step_diff);
+  step_diff_raw_prev_ = static_cast<int16_t>(step_in);
+  return output;
+}
+
+/**
+ * Function called when the filter is enabled
+ *
+ * Output smoothed differences.
+ * The smaller the difference, the longer the sampling time is stretched and smoothed.
+ * If the latest difference is greater than max_step_diff, no smoothing is performed.
+ * Smoothing is performed for max_sampling_time seconds when the latest difference is 0.
+ *
+ * @step_in Enter the current counter value
+ * @reset Used for initialization.
+ * @return Returns the difference from the previous value.
+ */
+double VescStepDifference::stepDifferenceVariableWindow(const double step_in, bool reset)
+{
+  if (reset)
+  {
+    for (int i = 0; i < step_diff_log_.size(); i++)
+    {
+      step_diff_log_[i] = static_cast<int16_t>(step_in);
+    }
+    return 0.0;
+  }
+  // Increment buffer
+  for (int i = step_diff_log_.size() - 1; i > 0; i--)
+  {
+    step_diff_log_[i] = step_diff_log_[i - 1];
+  }
+  step_diff_log_[0] = static_cast<int16_t>(step_in);
+  // Calculate window size
+  int latest_step_diff = abs(step_diff_log_[0] - step_diff_log_[1]);
+  int window_size = step_diff_vw_max_window_size_;
+  if (latest_step_diff > step_diff_vw_max_step_ - 1)
+  {
+    window_size = 1;
+  }
+  else if (latest_step_diff > 0)
+  {
+    window_size = step_diff_vw_max_window_size_ *
+                  (1.0 - static_cast<double>(latest_step_diff) / static_cast<double>(step_diff_vw_max_step_));
```

Comment:
Your understanding is correct, the larger the latest_step_diff, the more recent the step_diff_log to be extracted.
If minus 1, the sampling period is the same as for 1 because abs is taken.
If the step difference is less than 1, there is only 0.
If the difference of steps is 0, it takes the maximum sampling period.

Reply:
I will rename step_diff_log_ to step_input_queue_ because the name is inappropriate

Reply:
ok, glad to see i understood correctly